### PR TITLE
feat(plugins): adds endpoint to return list of plugins with projectList

### PR DESCRIPTION
**Pull request by scefali**
This PR adds a new endpoint which returns a list of all plugins (from the `PluginSerializer`) along with a field called `projectList` which contain information about the projects for which the plugin is configured (or simply enabled if no config is necessary for that plugin). Each entry in the list has the following fields:

- projectId
- projectSlug
- projectName
- enabled (whether the plugin is enabled for that project)
- configured (whether the plugin is configured for that project

The route for this endpoint is `/api/0/organizations/:org_slug/plugins/configs/`. This endpoint will be used in the integration directory home page which shows a list of all plugins along with a configuration count.

Note that this endpoint is somewhat similar to the existing endpoint `OrganizationPluginsEndpoint`. This endpoint will eventually replace the existing one once we have removed the old integrations page and replaced it with the integration directory. We can then use the `/api/0/organizations/:org_slug/plugins/` for this route.

Also note that I have included some fields that might not be needed like the `projectName` and `configured`. We can remove them if we fubd that the UI does not need them.

Note that I was somewhat concerned about the performance of this endpoint since we could query many projects so I tested it out on stage with an org that has in the top 1% in the number of projects. Our APM is pretty cool and useful! Here is the link: https://bit.ly/2t7BUvQ

Screen Shot 2020-01-29 at 10 18 53 AM

We are missing some of the instrumentation but it turns out the fetch for the projects only takes 10ms! For the large org I tested it with, the full request takes somewhere between 300ms and 750ms which is pretty decent for that org (`/organizations/:org_slug/?detailed=1` which called on all settings pages often takes 10+ seconds to complete). With other orgs the call is much faster as you would expect (with Sentry it's often about 150 ms)

### CODE REVIEW

File: src/sentry/runner/initializer.py [hunk 1]
```diff
@@ -60,6 +60,23 @@ def register_plugins(settings, raise_on_plugin_load_failure=False):
             pass
 
 
+def unregister_plugins(settings):
```
Comment (scefali) [inline src/sentry/runner/initializer.py]: @wedamija Would like to get your eyes on how I am doing the register/unregister of plugins since this is new.

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+class OrganizationPluginsConfigsEndpoint(OrganizationEndpoint):
+    def get(self, request, organization):
+        all_plugins = dict([(p.slug, p) for p in plugins.all()])
+
+        # find all the keys that will tell us if a plugin is enabled OR configured
+        keys_to_check = []
+        for slug, plugin in six.iteritems(all_plugins):
+            keys_to_check.append("%s:enabled" % slug)
+            if plugin.required_field:
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/organization_plugins_configs.py]: where is the `required_field` getting set again?

Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: @MeredithAnya They are defined in the `plugin.py` files where we sub-class the plugin. Ex:
```
class TrelloPlugin(CorePluginMixin, IssuePlugin2):
    description = "Create Trello cards on issues"
    slug = "trello"
    title = "Trello"
    conf_title = title
    conf_key = "trello"
    auth_provider = None
    resource_links = [("Trello Setup Instructions", SETUP_URL)] + CorePluginMixin.resource_links
    required_field = "key"

```

Comment (MeredithAnya) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: ahh okays cool - I haven't pulled from master in a while so I couldn't find when searching in my editor lol

Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: @MeredithAnya Hahah yea, that change was merged today :)

**Comment by scefali** [comment]
Closing as I need to make changes

File: src/sentry/runner/initializer.py (hunk 1 above)
Comment (scefali) [reply src/sentry/runner/initializer.py]: @wedamija I think this PR is in a reviewable state again. Might be worth checking out the actual implementation as well since this route became a bit heavier with my latest changes.

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        # plugins that can't be configured shouldn't be shown
+        all_plugins = dict([(p.slug, p) for p in plugins.plugins_with_configuration()])
+
+        # can specify specific plugins to load
+        if "plugins" in request.GET:
+            desired_plugins = request.GET.getlist("plugins")
+        else:
+            desired_plugins = all_plugins.keys()
+
+        # find all the keys that will tell us if a plugin is enabled OR configured
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/organization_plugins_configs.py]: I feel like maybe this could be a little clearer (this might be too wordy i dk up to you)
```python
# `keys_to_check` are the ProjectOption keys that tell us if a plugin is enabled (e.g. `plugin:enabled`) or are 
# configured properly, meaning they have the required information - plugin.required_field - needed for the
# plugin to work (e.g.`plugin:auth_key`)
```

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            if plugin.required_field:
+                keys_to_check.append("%s:%s" % (slug, plugin.required_field))
+
+        # Get all the project options for org that have truthy values
+        project_options = ProjectOption.objects.filter(
+            key__in=keys_to_check, project__organization=organization
+        ).exclude(value__in=[False, ""])
+
+        # This map stores info about whether a plugin is configured and/or enabled
+        # the first key is the plugin slug
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/organization_plugins_configs.py]: I think maybe an example of the structure be easier - is this right?
```python
{
  "plugin_slug": {
      "project_id": {
           "enabled": True,
           "configure": False,
       },
       ...
   },
}
```

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+            # first add to the set of all projects by plugin
+            info_by_plugin_project.setdefault(slug, {}).setdefault(
+                project_id, {"enabled": False, "configured": False}
+            )
+
+            # next check if enabled
+            if field == "enabled":
+                info_by_plugin_project[slug][project_id]["enabled"] = True
+            # if the projectoption is not the enable field, it's connfiguration field
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/organization_plugins_configs.py]: small typo `connfiguration` -> `configuration`

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import six
+
+from sentry.plugins.base import plugins
+from sentry.api.bases.organization import OrganizationEndpoint
+from sentry.api.serializers import serialize
+from sentry.api.serializers.models.plugin import PluginSerializer
+from sentry.models import ProjectOption, Project
+
+
+# This endpoint is similar to OrganizationPluginsEndpoint
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: @MeredithAnya Thanks! Appreciate that you wrote the doc comment for me :)

File: src/sentry/runner/initializer.py [hunk 7]
```diff
@@ -60,6 +60,19 @@ def register_plugins(settings, raise_on_plugin_load_failure=False):
...
```
Comment (wedamija) [inline src/sentry/runner/initializer.py]: What is settings used for here?

File: tests/sentry/api/endpoints/test_organization_plugins_configs.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+from sentry.plugins.base import plugins
+from sentry.testutils import APITestCase
+from sentry.runner.initializer import register_plugins, unregister_plugins
+
+
+class OrganizationPluginsTest(APITestCase):
+    def setUp(self):
+
+        register_plugins(settings, raise_on_plugin_load_failure=True)
```
Comment (wedamija) [inline tests/sentry/api/endpoints/test_organization_plugins_configs.py]: Does calling this multiple times result in the `SENTRY_DEFAULT_INTEGRATIONS ` being registered multiple times and never unregistered?

File: src/sentry/runner/initializer.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/runner/initializer.py]: I'm wondering if it's worth doing something else here, rather than registering/unregistering plugins at the global level, which I can imagine might fail in various ways down the line.

One option could be to do something like
`mock.patch("sentry.plugins.base.plugins", new=PluginManager())` and then you can just register your changes as usual, and they'll be automatically rolled back once the mock unregisters. You may have to mock `register` and `unregister` too.

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+- similar to the `OrganizationPluginsEndpoint`, and can eventually replace it
+
+:qparam plugins array[string]: an optional list of plugin ids (slugs) if you want specific plugins.
+                               If not set, will return configurations for all plugins.
+"""
+
+
+class OrganizationPluginsConfigsEndpoint(OrganizationEndpoint):
+    def get(self, request, organization):
+
```
Comment (MeredithAnya) [inline src/sentry/api/endpoints/organization_plugins_configs.py]: the docstrings usually go right under the request method just fyi

Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: whoops... :)

File: tests/sentry/api/endpoints/test_organization_plugins_configs.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+from sentry.utils.compat import mock
+from sentry.plugins.base import plugins
+from sentry.testutils import APITestCase
+from sentry.runner.initializer import register_plugins
+
+
+class OrganizationPluginsTest(APITestCase):
+    # create a copy of class_list for plugins from the initial value
+    @mock.patch.object(plugins, "class_list", new=copy(plugins.class_list))
```
Comment (wedamija) [inline tests/sentry/api/endpoints/test_organization_plugins_configs.py]: Thanks for this change, I think this is much cleaner

**Review by wedamija — COMMENTED** [review]
Looks good to me, just a couple thoughts, will leave it to @MeredithAnya to approve

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        # if no plugins were specified, grab all plugins but limit by those that have the ability to be configured
+        if not desired_plugins:
+            desired_plugins = list(plugins.plugin_that_can_be_configured())
+
+        # `keys_to_check` are the ProjectOption keys that tell us if a plugin is enabled (e.g. `plugin:enabled`) or are
+        # configured properly, meaning they have the required information - plugin.required_field - needed for the
+        # plugin to work (ex:`opsgenie:api_key`)
+        keys_to_check = []
+        for plugin in desired_plugins:
+            keys_to_check.append("%s:enabled" % plugin.slug)
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: @wedamija The only problem with that is we have two main base classes for Plugins for the `v1` plugins and `v2`. Each of those subclass `PluginConfigMixin` but the `slug` field is defined on the `v1` and v2` classes respectively (most fields come from those classes). So I would end up having to have two different but identical methods to construct these keys.

File: src/sentry/api/endpoints/organization_plugins_configs.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            if plugin.required_field:
+                keys_to_check.append("%s:%s" % (plugin.slug, plugin.required_field))
+
+        # Get all the project options for org that have truthy values
+        project_options = ProjectOption.objects.filter(
+            key__in=keys_to_check, project__organization=organization
+        ).exclude(value__in=[False, ""])
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: @wedamija It does work:
```
SELECT "sentry_projectoptions"."id", "sentry_projectoptions"."project_id", "sentry_projectoptions"."key", "sentry_projectoptions"."value" FROM "sentry_projectoptions" INNER JOIN "sentry_project" ON ("sentry_projectoptions"."project_id" = "sentry_project"."id") WHERE ("sentry_project"."organization_id" = 1 AND "sentry_projectoptions"."key" IN (phabricator:enabled, redmine:host, pivotal:enabled, opsgenie:api_key, amazon-sqs:enabled, vsts:enabled, slack:webhook, webhooks:enabled, amazon-sqs:queue_url, splunk:enabled, jira:username, trello:enabled, pushover:apikey, splunk:instance, teamwork:url, clubhouse:token, webhooks:urls, pagerduty:service_key, bitbucket:repo, bitbucket:enabled, opsgenie:enabled, pagerduty:enabled, github:repo, asana:workspace, gitlab:gitlab_url, victorops:enabled, slack:enabled, pushover:enabled, victorops:api_key, gitlab:enabled, trello:key, vsts:instance, segment:write_key, clubhouse:enabled, pivotal:token, jira:enabled, twilio:enabled, phabricator:host, twilio:account_sid, segment:enabled, github:enabled, redmine:enabled, teamwork:enabled, asana:enabled) AND NOT ("sentry_projectoptions"."value" IN (gAKJLg==, gAJYAAAAAC4=)))
```

**Review by MeredithAnya — APPROVED** [review]

File: src/sentry/api/endpoints/organization_plugins_configs.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/api/endpoints/organization_plugins_configs.py]: lol, I like `NOT ("sentry_projectoptions"."value" IN ('gAKJLg==', 'gAJYAAAAAC4='))`. Looks fine though, I guess those are just the pickled values.

**Review by wedamija — APPROVED** [review]
lgtm

**Comment by billyvg** [comment]
@scefali I bet if you looked at the frontend instrumentation you would get very different results

**Comment by scefali** [comment]
@billyvg what do you mean by that?

**Comment by billyvg** [comment]
@scefali The query itself can be fast, but everything else that happens between the client/browser and query can slow it down (e.g. serialization)

File: conftest.py [hunk 13]
```diff
@@ -18,11 +18,8 @@ def pytest_configure(config):
     # being used
     warnings.filterwarnings("error", "", Warning, r"^(?!(|kombu|raven|sentry))")
 
-    # if we are running any tests for plugins, we need to make sure we install them first
-    # (for `py.test --version` or `py.test --help`, there are no files to test)
-    test_targets = config.getoption("file_or_dir")
-    if test_targets and any("tests/sentry_plugins" in s for s in test_targets):
-        install_sentry_plugins()
+    # always install plugins for the tests
+    install_sentry_plugins()
```
Comment (joshuarli) [inline conftest.py]: nice
